populator-machinery/controller.go

```diff
@@ -527,13 +629,51 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
 	}
 
 	// Look for the populator pod
-	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
-	c.addNotification(key, "pod", c.populatorNamespace, podName)
 	var pod *corev1.Pod
-	pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return err
+	podName := fmt.Sprintf("%s-%s", populatorPodPrefix, pvc.UID)
+	if c.populatorArgs != nil {
+		c.addNotification(key, "pod", c.populatorNamespace, podName)
+		pod, err = c.podLister.Pods(c.populatorNamespace).Get(podName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	// TODO: Clean up StroageClass prime when the original StorageClass gets deleted
+	// TODO: Add finializer to ensure user doesn't delete the SCPrime
+	// If populate data with provider implementation, and orginal StorageClass's VolumeBindingMode is VolumeBindingWaitForFirstConsumer,
+	// create a StorageClass with VolumeBindingImmediate for pvcPrime. Otherwise, pv will not get provisioned since we will not create the
+	// populator pod
+	storageClassPrimeName := *pvc.Spec.StorageClassName
+	if c.populateFn != nil && storageClass.VolumeBindingMode != nil && storagev1.VolumeBindingWaitForFirstConsumer == *storageClass.VolumeBindingMode {
+		storageClassPrimeName = fmt.Sprintf("%s-%s", populatorStorageClassPrefix, *pvc.Spec.StorageClassName)
+		scPrime, err := c.scLister.Get(storageClassPrimeName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			bm := storagev1.VolumeBindingImmediate
```

Comment:
I've been slightly wary of doing this automatically because it could cause unschedulable pods if the pvc gets provisioned in the wrong zone. I am thinking that we may want to defer this logic to some new prepareProvisionFn instead. Then the populator could decide if it needs to do this + add some AllowedTopolgoies to the StorageClass. WDYT? also cc @saikat-royc for thoughts.

Reply:
agree. this would be error prone for PD (we had a round of discussion around it). 
So if I understand any storage class prime management would be left to the provider specific impl

Reply:
Yes, that's correct

Reply:
Discussed offline, removed the StorageClass prime